libcontainer/rootfs_linux.go

```diff
@@ -747,12 +740,28 @@ func checkProcMount(rootfs, dest string, m mountEntry) error {
 		return nil
 	}
 	if path == "." {
-		// only allow a mount on-top of proc if it's source is "proc"
-		st, err := m.srcStatfs()
-		if err != nil {
-			return err
-		}
-		if st.Type == unix.PROC_SUPER_MAGIC {
+		// Onlu allow bind-mounts on top of /proc, and only if the source is a
+		// procfs mount.
+		if m.IsBind() {
```

Comment:
nit: I think this looks cleaner if we first return on the cases we are not interested on. Like: `if m.Device == "proc" && !m.IsBind()...; else if !m.IsBind()` and then just continue with this without identation.

Reply:
My worry was that it's very easy to miss that you need to be careful about how you check the type and `m.IsBind`. Also, the structure makes it clearer that it's fail-closed (this is how the old version worked too, and I didn't feel it was necessary to change it).